app/src/main/java/com/example/drawingactivity/DialogUtils.kt

```diff
@@ -137,6 +135,223 @@ object DialogUtils {
+            satText.forceText(displaySat.toString())
+            valText.forceText(displayVal.toString())
+            redText.forceText(colorState.red.toString())
+            greenText.forceText(colorState.green.toString())
+            blueText.forceText(colorState.blue.toString())
+            huePicker.setHue(colorState.hue)
+            gradientPicker.setHue(colorState.hue)
+            gradientPicker.setSaturationAndValue(colorState.saturation, colorState.value)
+            topColorSquare.setBackgroundColor(parseColor(colorState.hexCode))
+        }
+
+        /**
+         * Clears focus on whatever view the user has currently selected.
+         * Currently only called when the user interacts with the Gradient or Hue Slider.
+         */
+        fun clearFocus() {
+            hexText.clearFocus()
+            hueText.clearFocus()
+            satText.clearFocus()
+            valText.clearFocus()
+            redText.clearFocus()
+            greenText.clearFocus()
+            blueText.clearFocus()
+        }
+
+        val colorState = ColorUtils.calculateColorWithHsv(previousColor)
+        forceViewUpdates(colorState)
+        bottomColorSquare.setBackgroundColor(parseColor(colorState.hexCode))
+        huePicker?.apply {
+            onHueChanged = { it ->
+                if (satText.text.toString().toIntOrNull() != null && valText.text.toString()
+                        .toIntOrNull() != null
+                ) {
+                    clearFocus()
+                    val calculatedColorState = ColorUtils.calculateColorWithHsv(
+                        floatArrayOf(
+                            it.toFloat(),
+                            (satText.text.toString().toFloat() / 100),
+                            (valText.text.toString().toFloat() / 100)
+                        )
+                    )
+                    gradientPicker.setHue(it)
+                    forceViewUpdates(calculatedColorState)
+                }
+                onSliderClicked = { viewClicked ->
+                    sheet.behavior.isDraggable = !viewClicked
+                }
+            }
+        }
+        gradientPicker?.apply {
+            onSatValChanged = { saturation, value ->
+                if (hueText.text.toString().toIntOrNull() != null) {
+                    clearFocus()
+                    val calculatedColorState = ColorUtils.calculateColorWithHsv(
+                        floatArrayOf(
+                            (hueText.text.toString().toFloat()), saturation, 1f - value
+                        )
+                    )
+                    forceViewUpdates(calculatedColorState)
+                }
+            }
+            onGradientClicked = { viewClicked ->
+                sheet.behavior.isDraggable = !viewClicked
+            }
+            hexText.addTextChangedListener {
+                if (ColorUtils.isHexadecimal(it.toString())) {
+                    val color = parseColor("#${it.toString()}")
```

Comment:
AAAAAAAAA won't be kind to this will it?

Reply:
it would not, using a better checking method